### PR TITLE
[RFC] audit: unify build package

### PR DESCRIPTION
**Pull request by feckert**
@dangowrt @flyn-org @aparcar 

I wanted to use the audit to track changes in configuration files and some other events on the system.
That's why I started to work with audit.

I have now noticed a few strange things that I would like to solve and discuss in this pull request with you again.
The way it looks, the @dangowrt libaudit is built in the openwrt core and the other part of audit @aparcar in the package feed.
Both use the same build directory!
https://github.com/openwrt/packages/tree/master/utils/audit
I saw that this was already in one package together before.

But I think that this is not good, especially if there are version bumps that does not happen at the same time!

And then I noticed that the auditd does not have syscall tables for **all** architectures.
https://github.com/linux-audit/audit-userspace/pulls

So we should only build audit for this architectures.
https://github.com/openwrt/openwrt/compare/master...TDT-AG:pr/20200503-audit?expand=1#diff-1f32d39f9d0396e56263fac69e37d243910bbc377250f2a9bbcfe74ac684863fR32 that are supported.
For now only aarch64 || arm || powerpc64 || i386 || x86_64 are supported.

The package would be best placed in the package feed, but we have a [host build dependency](https://github.com/TDT-AG/openwrt/blob/master/package/libs/libsemanage/Makefile#L21) for libsemanage in the core.
So the rest of the compile package should also go into the openwrt core or we have to move the dependency also to the package feed.

I am also in the process of applying another pullrequest that supports the mips architecture, but there are still problems.
https://github.com/linux-audit/audit-userspace/pull/115
https://github.com/linux-audit/audit-userspace/pull/198

### CODE REVIEW

**Comment by flyn-org** [comment]
We went through several iterations as we added SELinux support to OpenWrt with the aim of maximally placing things in the packages repository while acknowledging that certain core pieces had to be in the main repository. There is a record of this at https://github.com/openwrt/packages/pull/10664 and https://github.com/openwrt/openwrt/pull/3207. LWN also wrote an article about SELinux on OpenWrt that captures the work we did: https://lwn.net/Articles/832876/. None of this directly answers your questions, but it should give you some necessary context.

**Comment by feckert** [comment]
@flyn-org thanks for hint.

I didn't realise that all this had already been discussed.
Independently of this, however, we should then check in the package feed whether the version of audit and libaudit fit together. We should also limit the architectures `aarch64 || arm || powerpc64 || i386 || x86_64`. The arch mips is not supported.

File: package/utils/audit/Makefile [hunk 1]
```diff
@@ -0,0 +1,184 @@
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
```
Comment (adschm) [inline package/utils/audit/Makefile]: Please use an SPDX license identifier instead.

**Comment by adschm** [comment]
All commits lack commit messages. While I understand this is RFC, you will require them eventually and it might even help reviewers at this stage.

**Comment by feckert** [comment]
@adschm thanks for your reply.
I wanted to use the RFC to initiate a discussion among the participants of the package maitainers.

@flyn-org Has already explained and pointed out to me how this all came about while selinux was being integrated into openwrt.

If there is a decision, then I will of course add the commit messages.
But so far it is not clear what is the right way to fix the problem I have found (Or is it a problem at all?)?

1. Leave it as it is (not recommended in my opinion)
2. Integrate everything into openwrt core (as written here in this pull request)
3. Move everything into the package feed.

The libaudit and audit packages are built from the same source code. And share that too during build!
From my point of view it is wrong.
Because it can lead to a version confusion.
Also, the packages are maintained in two different repos ([openwrt core](https://github.com/openwrt/openwrt/blob/master/package/libs/libaudit/Makefile) and [package feed](https://github.com/openwrt/packages/blob/master/utils/audit/Makefile)).

I have always understood this to mean that several install packages are to be built from one source package. That is not the case here.

Auditd is also not supported on every architecture. To be precise, we are talking about MIPS.
It is built, but the syscall tables are empty because there is [no support for this architecture](https://github.com/linux-audit/audit-userspace/pull/115) only a pullrequest to add this was staged some time ago.

File: package/utils/audit/Makefile (hunk 1 above)
Comment (feckert) [reply package/utils/audit/Makefile]: I have not changed that. For now, I have merged everything from the openwrt core and the package feed. To make it more comprehensible, I would have adjusted the things that are not right, e.g. SPDX and so on, only later.

**Comment by aparcar** [comment]
> The libaudit and audit packages are built from the same source code. And share that too during build!

Can't we just make them use different build directories and add a limitation to supported architectures? That seems easier than moving packages around.

**Comment by aparcar** [comment]
@feckert I'll test this locally, overall I think the idea is good and this is the proper fix. Please add commit messages explaining every step. Since big chunks of libaudit and audit are the same you could squash them as `audit: add package and merge with libaudit`.

**Comment by aparcar** [comment]
Additionally please create a PR at packages.git to remove `audit` over there and link the PR here, so it happens all at once.

File: package/utils/audit/Makefile [hunk 2]
```diff
@@ -105,5 +154,31 @@ define Package/libaudit/install
 	$(CP) $(PKG_INSTALL_DIR)/etc/libaudit.conf $(1)/etc/
 endef
 
+define Package/libauparse/install
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libauparse.so.* $(1)/usr/lib/
```
Comment (aparcar) [inline package/utils/audit/Makefile:159]: Did you test this? Above at libaudit/install should be more specific or you get the following error:

```
Collected errors:
 * check_data_file_clashes: Package libauparse wants to install file /home/user/src/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/root-mediatek/usr/lib/libauparse.so.0
	But that file is already provided by package  * libaudit
 * check_data_file_clashes: Package libauparse wants to install file /home/user/src/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/root-mediatek/usr/lib/libauparse.so.0.0.0
	But that file is already provided by package  * libaudit
 * opkg_install_cmd: Cannot install package libauparse.
 * check_data_file_clashes: Package libauparse wants to install file /home/user/src/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/root-mediatek/usr/lib/libauparse.so.0
	But that file is already provided by package  * libaudit
 * check_data_file_clashes: Package libauparse wants to install file /home/user/src/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/root-mediatek/usr/lib/libauparse.so.0.0.0
	But that file is already provided by package  * libaudit
```

**Review by aparcar — CHANGES_REQUESTED** [review]

File: package/utils/audit/Makefile [hunk 3]
```diff
@@ -28,14 +26,68 @@ PKG_INSTALL:=1
...
+$(call Package/audit/Default/description)
+  This package contains the audit utilities.
+endef
+
+define Package/audit
+$(call Package/audit/Default)
+  SECTION:=utils
+  CATEGORY:=Utilities
+  TITLE+= (daemon)
+  DEPENDS+= +libaudit +libauparse +audit-utils +libev +@KERNEL_AUDIT
```
Comment (aparcar) [inline package/utils/audit/Makefile:85]: This seem to depend on `libev` while we only have `libev2` in openwrt.git.

Comment (aparcar) [reply package/utils/audit/Makefile:85]: I may mix this up and libev is unrelated to libevent2, in that case we have to migrate the package as well.

**Review by aparcar — CHANGES_REQUESTED** [review]

File: package/utils/audit/Makefile (hunk 2 above)
Comment (feckert) [reply package/utils/audit/Makefile:159]: Good point. I have only imported this and have not changed the make source.
However, I have not installed it via opkg.
I will fix it

File: package/utils/audit/Makefile (hunk 3 above)
Comment (feckert) [reply package/utils/audit/Makefile:85]: This should be the https://github.com/openwrt/packages/blob/master/libs/libev/Makefile

I have also read in in the history that audit does have its own libev implementation if there is no libev found.
https://github.com/openwrt/packages/commit/55d588c9244f571315e942b9058c4a388dae4437#diff-1bb48de85a1017766fa1099efbdced6849929bdf1f42f256d83f8e03c78c8ed4

File: package/utils/audit/Makefile (hunk 2 above)
Comment (feckert) [reply package/utils/audit/Makefile:159]: fixed https://github.com/openwrt/openwrt/pull/4136/files#diff-1f32d39f9d0396e56263fac69e37d243910bbc377250f2a9bbcfe74ac684863fR152

File: package/utils/audit/Makefile (hunk 3 above)
Comment (aparcar) [reply package/utils/audit/Makefile]: @dangowrt can you comment here? I remember you weren't to keen to have libev in core?

**Comment by feckert** [comment]
This is replaced by the following pullrequest https://github.com/openwrt/openwrt/pull/4447.

**Pull request closed without merging**